vsce/Dockerfile

```diff
@@ -0,0 +1,28 @@
+# Start with an image that has Node.js.
+FROM node:16.13.0
+
+# create a WORKDIR to prevent
+# ERROR EPERM: operation not permitted
+WORKDIR /extension-publishing-helper-dir
+
+# copy current directory to image
+COPY . .
+
+# this is to avoid sh: 1: tsc: not found
+# DO NOT do "RUN npm install"! This
+# installs ALL dependencies, instead of
+# just the bare minimum necessary to
+# publish the extension. This could
+# easily increase the build time of this
+# step anywhere from 10x to 20x.
+RUN npm install -g typescript
```

Comment:
Actually, this doesn't work, as I noticed when editing this file recently.

Because the up-to two people who have used this `Dockerfile` to publish the extension, myself and possibly @DanBurton, already had the dependencies necessary to build the extension in our current directory, the preceding step in this `Dockerfile`, `COPY . .`, added those dependencies.

This is why `npm install` was not necessary: we **already had** those dependencies in our current directory and copied them over with `COPY . .`.

Well, if this works, why does it matter?

Eventually, we want to have our continuous integration (CI) automatically release the extension. _It_ won't have the dependencies already installed from previous releases, presumably, so this step would have failed; `COPY . .` wouldn't have fixed the problem like it did for us, without our knowing it.

A future version of this file fixes this by just doing `npm install` to install all of the required dependencies necessary to package our extension.

To confirm that this really was an issue, I reset our working directory to the commit where we merged this pull request into our main branch, deleted any pre-installed depedencies, and ran `KEY=ourSecretExtensionPublishingPersonalAccessToken make publish`.

This is the output...
```
Now attempting to publish version 0.15.11
of our extension...
VER=0.15.11 ./extensionPublishingHelper.sh
Building Docker image to publish
version 0.15.11 of our extension...
[+] Building 20.3s (10/10) FINISHED                                                                               
 => [internal] load build definition from Dockerfile                                                         0.1s
 => => transferring dockerfile: 38B                                                                          0.0s
 => [internal] load .dockerignore                                                                            0.1s
 => => transferring context: 2B                                                                              0.0s
 => [internal] load metadata for docker.io/library/node:16.13.0                                              5.1s
 => [internal] load build context                                                                            0.1s
 => => transferring context: 56.67kB                                                                         0.0s
 => [1/5] FROM docker.io/library/node:16.13.0@sha256:580a0850049c59a48f06090edd48c9f966c5e6572bbbabc369ba3e  0.0s
 => CACHED [2/5] WORKDIR /extension-publishing-helper-dir                                                    0.0s
 => [3/5] COPY . .                                                                                           0.2s
 => [4/5] RUN npm install -g typescript                                                                      4.2s
 => [5/5] RUN npm install -g vsce                                                                            9.8s 
 => exporting to image                                                                                       0.8s 
 => => exporting layers                                                                                      0.7s 
 => => writing image sha256:73dbbbeb171cb8cd7e0f2279e894fb579ff4fcf1958907617d692c76b9ba56d2                 0.0s 
 => => naming to docker.io/library/img                                                                       0.0s 
                                                                                                                  
Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
Running Docker image to publish
version 0.15.11 of our extension...
Executing prepublish script 'npm run vscode:prepublish'...

> reach-ide@1001.7.3 vscode:prepublish
> npm run compile


> reach-ide@1001.7.3 compile
> tsc -b

client/src/CommandsTreeDataProvider.ts(3,23): error TS2307: Cannot find module 'path' or its corresponding type declarations.
client/src/CommandsTreeDataProvider.ts(5,30): error TS2304: Cannot find name '__filename'.
client/src/CommandsTreeDataProvider.ts(6,34): error TS2304: Cannot find name '__filename'.
client/src/CommandsTreeDataProvider.ts(7,32): error TS2304: Cannot find name '__filename'.
client/src/CommandsTreeDataProvider.ts(8,31): error TS2304: Cannot find name '__filename'.
client/src/CommandsTreeDataProvider.ts(9,29): error TS2304: Cannot find name '__filename'.
client/src/CommandsTreeDataProvider.ts(45,4): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(4,30): error TS2307: Cannot find module 'child_process' or its corresponding type declarations.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(5,24): error TS2307: Cannot find module 'net' or its corresponding type declarations.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(9,26): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(15,26): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(29,26): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(32,27): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(46,62): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(46,99): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-jsonrpc/lib/node/main.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
client/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(6,57): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(6,88): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
client/node_modules/vscode-languageclient/lib/node/main.d.ts(2,21): error TS2307: Cannot find module 'child_process' or its corresponding type declarations.
client/node_modules/vscode-languageclient/lib/node/main.d.ts(57,13): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-languageclient/lib/node/main.d.ts(58,13): error TS2503: Cannot find namespace 'NodeJS'.
client/node_modules/vscode-languageclient/lib/node/main.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
client/src/extension.ts(13,23): error TS2307: Cannot find module 'path' or its corresponding type declarations.
client/src/extension.ts(15,22): error TS2307: Cannot find module 'child_process' or its corresponding type declarations.
client/src/extension.ts(27,18): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
client/src/extension.ts(33,12): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
client/src/extension.ts(34,13): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
client/src/test/helper.ts(7,23): error TS2307: Cannot find module 'path' or its corresponding type declarations.
client/src/test/helper.ts(35,22): error TS2304: Cannot find name '__dirname'.
client/src/test/completion.test.ts(7,25): error TS2307: Cannot find module 'assert' or its corresponding type declarations.
client/src/test/completion.test.ts(10,1): error TS2582: Cannot find name 'suite'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
client/src/test/completion.test.ts(13,2): error TS2582: Cannot find name 'test'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
client/src/test/diagnostics.test.ts(7,25): error TS2307: Cannot find module 'assert' or its corresponding type declarations.
client/src/test/diagnostics.test.ts(10,1): error TS2582: Cannot find name 'suite'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
client/src/test/diagnostics.test.ts(13,2): error TS2582: Cannot find name 'test'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
client/src/test/index.ts(5,23): error TS2307: Cannot find module 'path' or its corresponding type declarations.
client/src/test/index.ts(6,24): error TS2307: Cannot find module 'mocha' or its corresponding type declarations.
client/src/test/index.ts(17,20): error TS2304: Cannot find name '__dirname'.
client/node_modules/vscode-test/out/util.d.ts(2,24): error TS2307: Cannot find module 'https' or its corresponding type declarations.
client/node_modules/vscode-test/out/util.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
client/src/test/runTest.ts(5,23): error TS2307: Cannot find module 'path' or its corresponding type declarations.
client/src/test/runTest.ts(13,49): error TS2304: Cannot find name '__dirname'.
client/src/test/runTest.ts(17,43): error TS2304: Cannot find name '__dirname'.
client/src/test/runTest.ts(23,3): error TS2580: Cannot find name 'process'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/keywordCompletion.ts(12,5): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/keywordCompletion.ts(20,39): error TS2550: Property 'entries' does not exist on type 'ObjectConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2017' or later.
server/src/keywordCompletion.ts(25,7): error TS7053: Element implicitly has an 'any' type because expression of type 'any' can't be used to index type 'typeof CompletionItemKind'.
server/src/mapKeywordsWithAPeriodToAKeywordList.ts(12,5): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(4,30): error TS2307: Cannot find module 'child_process' or its corresponding type declarations.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(5,24): error TS2307: Cannot find module 'net' or its corresponding type declarations.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(9,26): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(15,26): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(29,26): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(32,27): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(46,62): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(46,99): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-jsonrpc/lib/node/main.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
server/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(6,57): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(6,88): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver-protocol/lib/node/main.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.
server/node_modules/vscode-languageserver/lib/node/main.d.ts(29,55): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver/lib/node/main.d.ts(29,92): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver/lib/node/main.d.ts(53,250): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver/lib/node/main.d.ts(53,287): error TS2503: Cannot find namespace 'NodeJS'.
server/node_modules/vscode-languageserver/lib/node/main.d.ts(2,23): error TS2688: Cannot find type definition file for 'node'.
server/src/server.ts(1,27): error TS2307: Cannot find module 'querystring' or its corresponding type declarations.
server/src/server.ts(2,21): error TS2307: Cannot find module 'url' or its corresponding type declarations.
server/src/server.ts(38,21): error TS2307: Cannot find module 'process' or its corresponding type declarations.
server/src/server.ts(61,68): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/server.ts(91,18): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/server.ts(93,44): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/server.ts(95,12): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/server.ts(96,12): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
server/src/server.ts(97,14): error TS2580: Cannot find name 'require'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.
 ERROR  npm failed with exit code 1
make: *** [Makefile:28: publish] Error 1
```

As one can see, there really was an error with keeping the `Dockerfile` this way; we just didn't notice it because we were unknowingly working around the issue with the preceding `COPY . .` step.